src/transformers/models/dab_detr/modeling_dab_detr.py

```diff
@@ -0,0 +1,2689 @@
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, encoder_states, all_attentions] if v is not None)
+        return BaseModelOutput(
+            last_hidden_state=hidden_states, hidden_states=encoder_states, attentions=all_attentions
+        )
+
+
+# Modified from transformers.models.conditional_detr.modeling_conditional_detr.ConditionalDetrDecoder with ConditionalDetr->DABDETR,Conditional DETR->DAB-DETR
+class DABDETRDecoder(DABDETRPreTrainedModel):
+    """
+    Transformer decoder consisting of *config.decoder_layers* layers. Each layer is a [`DABDETRDecoderLayer`].
+
+    The decoder updates the query embeddings through multiple self-attention and cross-attention layers.
+
+    Some small tweaks for DAB-DETR:
+
+    - object_queries and query_position_embeddings are added to the forward pass.
+    - if self.config.auxiliary_loss is set to True, also returns a stack of activations from all decoding layers.
+
+    Args:
+        config: DABDETRConfig
+    """
+
+    def __init__(self, config: DABDETRConfig):
+        super().__init__(config)
+        self.config = config
+        self.dropout = config.dropout
+        self.layerdrop = config.decoder_layerdrop
+        self.num_layers = config.decoder_layers
+        self.return_intermediate = True  # config.return_intermediate_decoder it's default true in the original code
+
+        self.layers = nn.ModuleList([DABDETRDecoderLayer(config) for _ in range(config.decoder_layers)])
+        # in DAB-DETR, the decoder uses layernorm after the last decoder layer output
+        self.layernorm = nn.LayerNorm(config.d_model)
+        d_model = config.d_model
+
+        # query_scale is the FFN applied on f to generate transformation T
+        assert config.query_scale_type in ["cond_elewise", "cond_scalar", "fix_elewise"]
+        self.query_scale_type = query_scale_type = config.query_scale_type
+        if query_scale_type == "cond_elewise":
+            self.query_scale = MLP(d_model, d_model, d_model, 2)
+        elif query_scale_type == "cond_scalar":
+            self.query_scale = MLP(d_model, d_model, 1, 2)
+        elif query_scale_type == "fix_elewise":
+            self.query_scale = nn.Embedding(config.decoder_layers, d_model)
+        else:
+            raise NotImplementedError("Unknown query_scale_type: {}".format(query_scale_type))
+
+        self.ref_point_head = MLP(config.decoder_query_dim // 2 * d_model, d_model, d_model, 2)
+
+        self.bbox_embed = None
+        self.d_model = d_model
+        self.decoder_modulate_hw_attn = config.decoder_modulate_hw_attn
+        self.decoder_bbox_embed_diff_each_layer = config.decoder_bbox_embed_diff_each_layer
+
+        if self.decoder_modulate_hw_attn:
+            self.ref_anchor_head = MLP(d_model, d_model, 2, 2)
+
+        if not config.keep_query_pos:
+            for layer_id in range(config.decoder_layers - 1):
+                self.layers[layer_id + 1].ca_qpos_proj = None
+
```

Comment:
can we instead use `keep_query_pos` in layers and not create this module if it's not needed?

Reply:
Moved to the layer init.